onnxruntime/core/graph/graph.cc

```diff
@@ -3971,21 +3972,16 @@ ONNX_NAMESPACE::GraphProto Graph::ToGraphProto() const {
   return result;
 }
 
-ONNX_NAMESPACE::GraphProto Graph::ToGraphProtoWithExternalInitializers(const std::string& external_file_name,
-                                                                       const PathString& destination_file_path,
+ONNX_NAMESPACE::GraphProto Graph::ToGraphProtoWithExternalInitializers(const std::filesystem::path& external_file_path,
+                                                                       const std::filesystem::path& model_file_path,
                                                                        size_t initializer_size_threshold) const {
   GraphProto result;
   ToGraphProtoInternal(result);
+  // If model_file_path is just a file name without a path separator, for example: "model.onnx". Its parent path could
+  // be empty. Else, save external data file in same directory as the model.
+  const std::filesystem::path modified_external_file_path = model_file_path.parent_path() / external_file_path;
```

Comment:
allowing a path for `external_file_path/name` changes the semantics a bit. how should a path with subdirectories or an absolute path be handled?

it might be better to keep `external_file_name` as a string if we only want to allow specification of a file name that should be in the model directory.

Reply:
So there are a few cases:
1. model_file_path is empty. In this case we will use external_file_path as the path for storing the tensors.
2. model_file_path is not empty, but it doesn't have a parent path.  For example, it is just: "model.onnx". We will do the same as the above. 
3. model_file_path is not empty and it has a parent path. We will concat the parent path and the supplied external_file_path as the path for storing the tensors.

I think the logic is good. Anything wrong with it?

Reply:
e.g., one to consider is if external_file_path is an absolute path. appending an absolute path to an existing path may replace the existing path.

we could possibly first check that `external_file_path` is a relative path with one component that is a filename. something like `external_file_path.has_filename() && external_file_path.filename() == external_file_path`

Reply:
If I remember correctly, ONNX doesn't allow arbitrary paths for the external tensors. What if we just throw an error in that case?

Reply:
sure, checking and throwing an error would be good

Reply:
hm, if we do allow a path with multiple elements for `external_file_path`, do we need consider how to handle `../../../../some_directory_outside_of_the_onnx_model_dir`?

Reply:
Maybe we should keep the code simple until there is such a need?

Reply:
yeah, I think there is no need to allow it now. would the check I proposed earlier work?

`external_file_path.has_filename() && external_file_path.filename() == external_file_path`

Reply:
The "external_file_path" could be something like "aaaa/bbb.txt" ?

Reply:
oh. the old name `external_file_name` had led me to assume this should be a file name.

I guess it only needs to be a path relative to the model directory:
https://github.com/onnx/onnx/blob/4c3ed5e08be75bbe1eeb6818e490b1b6a370183e/onnx/onnx.in.proto#L622-L623

so checking for `external_file_path` being relative should be good